File: gawk-server/wire/stripe.go
package wire

// Striped delivery for the datagram delta path (R30, docs/35).
//
// A striping viewer holds its primary subscribe session plus up to
// MaxStripeLegs additional "leg" sessions, each dialed with ?stripe=N&leg=j
// and carrying exactly the delta datagrams whose ordinal d satisfies
// d mod N == j (data chunk i has ordinal i; parity symbol r has ordinal
// chunkCount+r, so parity keeps its measured tail-of-burst position on every
// leg). The only new wire message is StripeState: the viewer telling the
// relay to stop sending delta datagrams on the primary while the legs carry
// them. Everything else — chunk and parity headers, keyframe streams, audio,
// control — is byte-identical to pre-R30; striping redistributes datagrams,
// it never rewrites them (docs/35 §5.2).

import "fmt"

const (
	// StripeStateSize is the exact size of a StripeState datagram:
	// version, type, flags, stripeN, reserved.
	StripeStateSize = 5

	// MaxStripeLegs bounds a viewer's stripe width. Evidence-bound, not
	// tuned: docs/34 finding 5 measured per-connection headroom at up to 4
	// simultaneous connections and no further, and 4 legs at the burst
	// target of 6 chunks cover every delta frame size observed (~23 chunks
	// max). A constant, not a knob (docs/35 §11) — revisit only with a new
	// finding-5-shaped measurement at higher N.
	MaxStripeLegs = 4

	// stripeFlagStriped is bit 0 of the StripeState flags byte: the viewer
	// has live legs and wants delta datagrams suppressed on this session.
	stripeFlagStriped = 0x01
)

// StripeState is the primary-suppression signal (R30, docs/35 §5.3):
// client→relay, datagram, valid only on an external datagram-delivery
// subscribe session that is not itself a leg.
//
// It is LEVEL state, not an edge: the viewer re-sends it at 1 Hz while
// striped (plus a short burst on the unstripe transition), and the relay
// expires the suppression if refreshes stop — so every lost-message state
// converges to duplicates, never to holes.
type StripeState struct {
	// Striped reports whether delta datagrams should be suppressed on the
	// session this message arrived on.
	Striped bool
	// StripeN is the viewer's current stripe width, informational (for
	// /statusz and metrics — the relay's own behavior never depends on it).
	// In [1, MaxStripeLegs] when Striped, 0 otherwise.
	StripeN uint8
}

// AppendStripeState appends a StripeState datagram to dst.
func AppendStripeState(dst []byte, s StripeState) ([]byte, error) {
	if err := validateStripeState(s); err != nil {
		return nil, err
	}
	var flags uint8
	if s.Striped {
		flags |= stripeFlagStriped
	}
	dst = append(dst, Version, TypeStripeState, flags, s.StripeN, 0)
	return dst, nil
}

// ParseStripeState parses a StripeState datagram. Strict: unknown flag bits
// are rejected rather than ignored — a future revision of this message would
// gate on a new RelayCapabilities bit, never on old relays guessing.
func ParseStripeState(b []byte) (StripeState, error) {
	if len(b) != StripeStateSize {
		return StripeState{}, fmt.Errorf("%w: %d bytes, want exactly %d",
			ErrShortDatagram, len(b), StripeStateSize)
	}
	if b[0] != Version {
		return StripeState{}, fmt.Errorf("%w: 0x%02x", ErrBadVersion, b[0])
	}
	if b[1] != TypeStripeState {
		return StripeState{}, fmt.Errorf("%w: got 0x%02x, want stripe state 0x%02x",
			ErrBadType, b[1], TypeStripeState)
	}
	if b[2]&^uint8(stripeFlagStriped) != 0 {
		return StripeState{}, fmt.Errorf("%w: unknown stripe flags 0x%02x", ErrBadType, b[2])
	}
	s := StripeState{Striped: b[2]&stripeFlagStriped != 0, StripeN: b[3]}
	if err := validateStripeState(s); err != nil {
		return StripeState{}, err
	}
	return s, nil
}

// validateStripeState holds the one shape rule for both directions: a striped
// state names its width, an unstriped one names none.
func validateStripeState(s StripeState) error {
	if s.Striped {
		if s.StripeN < 1 || s.StripeN > MaxStripeLegs {
			return fmt.Errorf("%w: stripeN %d, want [1, %d] while striped",
				ErrBadChunkCount, s.StripeN, MaxStripeLegs)
		}
		return nil
	}
	if s.StripeN != 0 {
		return fmt.Errorf("%w: stripeN %d, want 0 while not striped",
			ErrBadChunkCount, s.StripeN)
	}
	return nil
}

// StripeOrdinal returns the stripe ordinal of a delta datagram: data chunk i
// has ordinal i, parity symbol r over an n-chunk frame has ordinal n+r. Leg j
// of stripe N carries the datagrams with StripeOrdinal % N == j.
func StripeOrdinal(chunkIndex, chunkCount uint16, parityIndex int) uint32 {
	if parityIndex >= 0 {
		return uint32(chunkCount) + uint32(parityIndex)
	}
	return uint32(chunkIndex)
}

File: gawk-broadcast-windows/crates/wire/src/stripe.rs
//! Striped delivery (R30, docs/35), mirroring gawk-server/wire/stripe.go.
//! This producer neither sends nor receives StripeState — it is viewer↔relay —
//! but the vectors are mirrored by rule, exactly like the carrier framing.

use crate::error::WireError;
use crate::{TYPE_STRIPE_STATE, VERSION};

/// Exact size of a StripeState datagram.
pub const STRIPE_STATE_SIZE: usize = 5;

/// Bounds a viewer's stripe width. Evidence-bound, not tuned (docs/35 §11).
pub const MAX_STRIPE_LEGS: u8 = 4;

const STRIPE_FLAG_STRIPED: u8 = 0x01;

/// The primary-suppression signal: level state, re-sent at 1 Hz while striped.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct StripeState {
    /// Whether delta datagrams should be suppressed on the session this
    /// message arrived on.
    pub striped: bool,
    /// The viewer's current stripe width, informational. In
    /// `[1, MAX_STRIPE_LEGS]` when striped, 0 otherwise.
    pub stripe_n: u8,
}

fn validate(s: &StripeState) -> Result<(), WireError> {
    if s.striped {
        if s.stripe_n < 1 || s.stripe_n > MAX_STRIPE_LEGS {
            return Err(WireError::BadChunkCount {
                index: s.stripe_n.into(),
                count: MAX_STRIPE_LEGS.into(),
            });
        }
        return Ok(());
    }
    if s.stripe_n != 0 {
        return Err(WireError::BadChunkCount {
            index: s.stripe_n.into(),
            count: 0,
        });
    }
    Ok(())
}

/// Appends a StripeState datagram.
pub fn append_stripe_state(dst: &mut Vec<u8>, s: &StripeState) -> Result<(), WireError> {
    validate(s)?;
    let flags = if s.striped { STRIPE_FLAG_STRIPED } else { 0 };
    dst.extend_from_slice(&[VERSION, TYPE_STRIPE_STATE, flags, s.stripe_n, 0]);
    Ok(())
}

/// Parses a StripeState datagram. Strict: unknown flag bits are rejected —
/// a future revision gates on a new RelayCapabilities bit, never on old
/// parsers guessing.
pub fn parse_stripe_state(msg: &[u8]) -> Result<StripeState, WireError> {
    if msg.len() != STRIPE_STATE_SIZE {
        return Err(WireError::ShortDatagram {
            len: msg.len(),
            need: STRIPE_STATE_SIZE,
        });
    }
    if msg[0] != VERSION {
        return Err(WireError::BadVersion(msg[0]));
    }
    if msg[1] != TYPE_STRIPE_STATE {
        return Err(WireError::BadType {
            got: msg[1],
            want: TYPE_STRIPE_STATE,
        });
    }
    if msg[2] & !STRIPE_FLAG_STRIPED != 0 {
        return Err(WireError::UnknownStripeFlags(msg[2]));
    }
    let s = StripeState {
        striped: msg[2] & STRIPE_FLAG_STRIPED != 0,
        stripe_n: msg[3],
    };
    validate(&s)?;
    Ok(s)
}

/// The stripe ordinal of a delta datagram: data chunk i has ordinal i, parity
/// symbol r over an n-chunk frame has ordinal n+r. Leg j of stripe N carries
/// the datagrams with ordinal % N == j.
pub fn stripe_ordinal(chunk_index: u16, chunk_count: u16, parity_index: Option<u8>) -> u32 {
    match parity_index {
        Some(r) => u32::from(chunk_count) + u32::from(r),
        None => u32::from(chunk_index),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn stripe_state_shape_rules() {
        // Striped names its width…
        assert!(
            append_stripe_state(
                &mut Vec::new(),
                &StripeState {
                    striped: true,
                    stripe_n: 0
                }
            )
            .is_err()
        );
        assert!(
            append_stripe_state(
                &mut Vec::new(),
                &StripeState {
                    striped: true,
                    stripe_n: MAX_STRIPE_LEGS + 1
                }
            )
            .is_err()
        );
        // …an unstriped one names none.
        assert!(
            append_stripe_state(
                &mut Vec::new(),
                &StripeState {
                    striped: false,
                    stripe_n: 1
                }
            )
            .is_err()
        );

        let mut msg = Vec::new();
        append_stripe_state(
            &mut msg,
            &StripeState {
                striped: true,
                stripe_n: 3,
            },
        )
        .unwrap();
        assert_eq!(
            parse_stripe_state(&msg).unwrap(),
            StripeState {
                striped: true,
                stripe_n: 3
            }
        );

        // Unknown flag bits are rejected, never masked.
        let mut bad = msg.clone();
        bad[2] |= 0x02;
        assert_eq!(
            parse_stripe_state(&bad),
            Err(WireError::UnknownStripeFlags(0x03))
        );
    }

    #[test]
    fn ordinals_put_parity_after_data() {
        assert_eq!(stripe_ordinal(0, 9, None), 0);
        assert_eq!(stripe_ordinal(8, 9, None), 8);
        assert_eq!(stripe_ordinal(0, 9, Some(0)), 9);
        assert_eq!(stripe_ordinal(0, 9, Some(1)), 10);
    }
}
